server/src/main/java/org/opensearch/index/mapper/ConstantKeywordFieldMapper.java

```diff
@@ -0,0 +1,191 @@
+import org.opensearch.common.annotation.PublicApi;
+import org.opensearch.common.regex.Regex;
+import org.opensearch.index.fielddata.IndexFieldData;
+import org.opensearch.index.fielddata.plain.ConstantIndexFieldData;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.search.aggregations.support.CoreValuesSourceType;
+import org.opensearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * Index specific field mapper
+ *
+ * @opensearch.api
+ */
+@PublicApi(since = "2.14.0")
+public class ConstantKeywordFieldMapper extends ParametrizedFieldMapper {
+
+    public static final String CONTENT_TYPE = "constant_keyword";
+
+    private static final String valuePropertyName = "value";
+
+    /**
+     * A {@link Mapper.TypeParser} for the constant keyword field.
+     *
+     * @opensearch.internal
+     */
+    public static class TypeParser implements Mapper.TypeParser {
+        @Override
+        public Mapper.Builder parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            if (!node.containsKey(valuePropertyName)) {
+                throw new OpenSearchParseException("Field [" + name + "] is missing required parameter [value]");
+            }
+            Object value = node.remove(valuePropertyName);
+            if (!(value instanceof String)) {
+                throw new OpenSearchParseException("Field [" + name + "] is expected to be a string value");
+            }
+            return new Builder(name, (String) value);
+        }
+    }
+
+    private static ConstantKeywordFieldMapper toType(FieldMapper in) {
+        return (ConstantKeywordFieldMapper) in;
+    }
+
+    /**
+     * Builder for the binary field mapper
+     *
+     * @opensearch.internal
+     */
+    public static class Builder extends ParametrizedFieldMapper.Builder {
+
+        private final Parameter<String> value;
+
+        public Builder(String name, String value) {
+            super(name);
+            this.value = Parameter.stringParam(valuePropertyName, false, m -> toType(m).value, value);
+        }
+
+        @Override
+        public List<Parameter<?>> getParameters() {
+            return Arrays.asList(value);
+        }
+
+        @Override
+        public ConstantKeywordFieldMapper build(BuilderContext context) {
+            return new ConstantKeywordFieldMapper(
+                name,
+                new ConstantKeywordFieldMapper.ConstantKeywordFieldType(buildFullName(context), value.getValue()),
+                multiFieldsBuilder.build(this, context),
+                copyTo.build(),
+                this
+            );
+        }
+    }
+
+    /**
+     * Field type for Index field mapper
+     *
+     * @opensearch.internal
+     */
+    @PublicApi(since = "2.14.0")
```

Comment:
Looks like this one should not have the `@PublicApi` annotation.

We just added a check last week to check that the `@PublicApi` annotation only appears on methods with `public` or `protected` visibility. In this case, IMO, the package-private visibility makes sense.

Reply:
I actually added this because I got a CI error saying this class needs to have the publicapi annotation
https://github.com/opensearch-project/OpenSearch/actions/runs/8443709580/job/23127848902

```
error: The element org.opensearch.index.mapper.ConstantKeywordFieldMapper.ConstantKeywordFieldType is part of the public APIs but does not have public or protected visibility (referenced by org.opensearch.index.mapper.ConstantKeywordFieldMapper) 
error: The element org.opensearch.index.mapper.ConstantKeywordFieldMapper.ConstantKeywordFieldType is part of the public APIs but is not marked as @PublicApi, @ExperimentalApi or @DeprecatedApi (referenced by 
```

I will make this class public In that case

Reply:
@msfroh 
Did this 
CI is failing with a log similar to master on mar 28
https://github.com/opensearch-project/OpenSearch/actions/runs/8468949056/job/23203116869

Reply:
I've done a rebase onto `main`. I'll follow along to try to get the build to pass today.

Reply:
Finally! All checks passed! 🎉

Reply:
Thanks @msfroh